src/celttf/truetypefont.cpp

```diff
@@ -0,0 +1,628 @@
+ */
+int TextureFont::getWidth(const string& s) const
+{
+    int width = 0;
+    int len = s.length();
+    bool validChar = true;
+    int i = 0;
+
+    while (i < len && validChar)
+    {
+        wchar_t ch = 0;
+        validChar = UTF8Decode(s, i, ch);
+        if (!validChar)
+            break;
+
+        i += UTF8EncodedSize(ch);
+
+        auto& g = impl->getGlyph(ch, L'?');
+        width += g.ax;
+    }
+
+    return width;
+}
+
+int TextureFont::getHeight() const
+{
+    return impl->m_maxAscent + impl->m_maxDescent;
+}
+
+int TextureFont::getMaxWidth() const
+{
+    return impl->m_maxWidth;
+}
+
+int TextureFont::getMaxAscent() const
+{
+    return impl->m_maxAscent;
+}
+
+void TextureFont::setMaxAscent(int _maxAscent)
+{
+    impl->m_maxAscent = _maxAscent;
+}
+
+int TextureFont::getMaxDescent() const
+{
+    return impl->m_maxDescent;
+}
+
+void TextureFont::setMaxDescent(int _maxDescent)
+{
+    impl->m_maxDescent = _maxDescent;
+}
+
+int TextureFont::getTextureName() const
+{
+    return impl->m_texName;
+}
+
+void TextureFont::bind()
+{
+    if (impl->m_texName != 0)
+        glBindTexture(GL_TEXTURE_2D, impl->m_texName);
+}
+
+short TextureFont::getAdvance(wchar_t ch) const
+{
+    auto& g = impl->getGlyph(ch, L'?');
+    return g.ax;
+}
+
+bool TextureFont::buildTexture()
+{
+    return true;
+}
+
+TextureFont* TextureFont::load(const fs::path &path, int size, int dpi)
+{
+    FT_Face face;
+
+    if (FT_New_Face(ft, path.string().c_str(), 0, &face) != 0)
+    {
+        fmt::fprintf(cerr, "Could not open font %s\n", path);
+        return nullptr;
+    }
+
+    if (!FT_IS_SCALABLE(face))
+    {
+        fmt::fprintf(cerr, "Font is not scalable: %s\n", path);
+        return nullptr;
+    }
+
+    if (FT_Set_Char_Size(face, 0, size << 6, dpi, dpi) != 0)
+    {
+        fmt::fprintf(cerr, "Could not font size %i\n", size);
+        return nullptr;
+    }
+
+    auto* font = new TextureFont();
+    font->impl->m_face = face;
+
+    if (!font->impl->buildAtlas())
+        return nullptr;
+
+    font->setMaxAscent(face->size->metrics.ascender >> 6);
+    font->setMaxDescent(-face->size->metrics.descender >> 6);
+
+    return font;
+}
+
+// temporary while no fontconfig support
+static fs::path ParseFontName(const fs::path &filename, int &size)
+{
+    auto fn = filename.string();
+    auto pos = fn.rfind(',');
+    if (pos != string::npos)
+    {
+        size = (int) stof(fn.substr(pos + 1));
+        return fn.substr(0, pos);
+    }
+    else
+    {
+        size = 12;
+        return filename;
+    }
+}
+
+TextureFont* LoadTextureFont(const fs::path &filename, int size, int dpi)
+{
+    if (ft == nullptr)
+    {
+        if (FT_Init_FreeType(&ft))
+        {
+            cerr << "Could not init freetype library\n";
+            return nullptr;
+        }
+    }
+
+    int psize = 0;
+    auto nameonly = ParseFontName(filename, psize);
+    return TextureFont::load(fs::path(CONFIG_DATA_DIR) / nameonly, psize, dpi);
```

Comment:
CONFIG_DATA_DIR might not be available for non-qt ones, I think we should just leave it blank, since we already changed current directory to CONFIG_DATA_DIR before performing any operation